### DISCUSSION TITLE
Several questions

### DISCUSSION THREAD

**Discussion by Darkloke** [oskvr37/tiddl#46] [Q&A] [answered]
A few suggestions or may be questions, if you don't mind. I am just trying to find the working alternative for tidal-dl by yarozz and so far all tools i tried have flaws (from my POV) ofc.

What is the Tidal music quality status for now? As i understand they changed something this summer and many downloaders just didnt follow that changes. For example streamripper has the following:
# 0: 256kbps AAC, 1: 320kbps AAC, 2: 16/44.1 "HiFi" FLAC, 3: 24/44.1 "MQA" FLAC
while tidal-dl-ng has this:
![image](https://github.com/user-attachments/assets/b531bf32-72ec-43d6-94f1-86d6b23ce62b)

How is your tool separate them?

Your tool only downloads in m4a? No FLACS or i am missing something?

As for feature requests - it would be nice to have some mechanism to track and redownload "failed" tracks. So for example if some track wasnt download for some reason (i.e. connection drop), a tool could easily redownload it. Nice idea was implemented in streamrip - it has a database for successful and failed track, so using one command 'repair' you could redownload all failed tracks.

What about parallel/multithreaded downloads?

**Comment by oskvr37** [comment, marked as answer]
TIDDL is using 4 qualities `low normal high master` which corespond to qualities available on Tidal:

![image](https://github.com/user-attachments/assets/ccc3904c-6aab-4c4e-a95e-06325a55f11f)

There is no 256kpbs on Tidal, I guess author of streamripper just made mistake.

There is FLAC, but first we need to download track in mp4a format as it is stored on Tidal in this format, then we save it as m4a file and add metadata and then we can convert it to FLAC.
Use option `-e flac` to set file extension to FLAC.

There will be database in the future to track downloaded and missing files.

Parallel downloading is not supported as you need first to make a request for every track (Tidal can detect this behavior) and we need to wait between every request, so making parallel downloading is pointless.
We have to download tracks one by one.

**Reply by Darkloke** [reply]
So a conversion to FLAC is made by the tool itself. Thats why i suppose other tools use ffmpeg for... But iirc FLAC conversion has some adjustable parameters like compression level - i suppose some defaults used here or this in theory could be handed over user control too?

Sad to hear about parallel downloading... But how this was done in yarozz's and streamripper? They somehow "cheated" the system? Or the whole speed of divided segments is still equal as it would download the single piece?

**Reply by oskvr37** [reply]
> i suppose some defaults used here or this in theory could be handed over user control too?

Yes, defaults are used and it could be controled by user.

> But how this was done in yarozz's and streamripper?

As I remember yaronzz had warning about multithreaded downloadind in his project.
I never used streamripper, you can give me a link to its repo, so I can check how they deal with multithreading.

TIDDL is using multithreading only when it is downloading master tracks, which are splitted in mp4a fragments.

**Reply by Darkloke** [reply]
> I never used streamripper, you can give me a link to its repo, so I can check how they deal with multithreading.

https://github.com/nathom/streamrip

**Reply by oskvr37** [reply]
I didnt look deep in its source code but maybe streamrip scrapes all tracks source first, and after scraping it downloads the tracks in parallel.
I can do some tests if this method will be faster and I think it will probably be.

**Reply by Darkloke** [reply]
Also a question about metadata. Lets take for example this album:
https://listen.tidal.com/album/88878173/credits
The tracks have "composer" data, but at downloaded flacs this field is empty. May be i am doing something wrong?

**Reply by oskvr37** [reply]
We would need to call `https://listen.tidal.com/v1/tracks/88878174/credits` endpoint to get track credits where we can get the composer data.
This feature is not implemented yet.

**Reply by Darkloke** [reply]
Ah, so you are already aware about this. Excellent, will be looking forward this feature, since for me metadata plays an important role.
